### PR TITLE
TLS Support for Bacula

### PR DESCRIPTION
**Pull request by RaHoni**
## Add support for TLS in the Bacula module

This PR adds support to configure TLS in Bacula. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaHoni** [comment]
@lovek323 and @proteansec is there anything that I should add or do I need to test something?

File: nixos/modules/services/backup/bacula.nix [hunk 1]
```diff
@@ -9,6 +9,21 @@ with lib;
...
+    concatStringsSep
+      "\n"
+      (
+      [''TLS Enable = yes;'']
+      ++ optional (tls_cfg.require != null) ''TLS Require = ${yes_no tls_cfg.require};''
+      ++ optional (tls_cfg.certificate != null) ''TLS Certificate = "${tls_cfg.certificate}";''
+      ++ [''TLS Key = "${tls_cfg.key}";'']
+      ++ optional (tls_cfg.verifyPeer != null) ''TLS Verify Peer = ${yes_no tls_cfg.verifyPeer};''
+      ++ optional (tls_cfg.allowedCN != { }) ''TLS Allowed CN = ${concatStringsSep " " (attrNames tls_cfg.allowedCN)};''
+      ++ optional (tls_cfg.caCertificateFile != null) ''TLS CA Certificate File = ${tls_cfg.caCertificateFile};''
```
Comment (JulianFP) [inline nixos/modules/services/backup/bacula.nix:23]:
```suggestion
      ++ optional (tls_cfg.allowedCN != { }) ''TLS Allowed CN = ${concatStringsSep " " tls_cfg.allowedCN};''
      ++ [''TLS CA Certificate File = "${tls_cfg.caCertificateFile}";'']
```

File: nixos/modules/services/backup/bacula.nix [hunk 2]
```diff
@@ -108,13 +128,94 @@ let
...
+          signed by a known-CA will be accepted. Additionally, the client's
+          X509 certificate Common Name must meet the value of the Address
+          directive. If the TLSAllowed CN configuration directive is used,
+          the client's x509 certificate Common Name must also correspond to
+          one of the CN specified in the TLS Allowed CN directive. This directive
+          is valid only for a server and not in client context. Standard from Bacula is 'yes'.
+         '';
+      };
+      allowedCN = mkOption {
+        type = types.attrsOf types.str;
```
Comment (JulianFP) [inline nixos/modules/services/backup/bacula.nix:190]:
```suggestion
        type = types.listOf types.str;
```

**Review by JulianFP — CHANGES_REQUESTED** [review]
Looks mostly good, however I noticed the following mistakes:

1. allowedCN should be an list of strings instead of an attrSet
2. CA Certificate File is a required option (no default defined), so it makes no sence to put it behing optional keyword
3. path of CA Certificate File was not in a string (quotes missing)

File: nixos/modules/services/backup/bacula.nix (hunk 1 above)
Comment (RaHoni) [reply nixos/modules/services/backup/bacula.nix:23]: As you can see at [Bacula Docs](https://www.bacula.org/11.0.x-manuals/en/main/Configuring_Director.html#SECTION0021200000000000000000) under `TLS Allowed CN = ` you are not supposed to add quotes to `CA Certificate File`.

Comment (JulianFP) [reply nixos/modules/services/backup/bacula.nix:23]: > As you can see at the [Bacula Docs](https://www.bacula.org/11.0.x-manuals/en/main/Configuring_Director.html#SECTION0021200000000000000000) under `TLS Allowed CN = ` you are not supposed to add quotes to `CA Certificate File`.

Correct me if I'm wrong, but while the example doesn't use quotes, `TLS CA Certificate File` accepts datatype `<Filename>` just like `TLS Key` and `TLS Certificate`. So shouldn't we treat these three options in the same way (meaning either quotes for all three of them or for non of them)?

Comment (RaHoni) [reply nixos/modules/services/backup/bacula.nix]: Yes you are Correct. But I am honestly not completely sure  witch is the better way. Since on the one hand the Documentation does not use Quotes and therefor it is the recommended style. On the other hand I am not sure about special characters like space in filenames.

Comment (RaHoni) [reply nixos/modules/services/backup/bacula.nix]: Just looked at the Documentation for the [Data Types](https://www.bacula.org/11.0.x-manuals/en/main/Customizing_Configuration_F.html#SECTION0020240000000000000000) 

> A directory is either a quoted or non-quoted string.

So lets use quoted strings just for easier handling of special characters.

**Review by JulianFP — CHANGES_REQUESTED** [review]
Besides from this one small mistake everything looks good to go. The indentation in the generated config file is all wrong, however this is irrelevant for bacula to work and was already the case before your PR. Since I also don't know an easy fix for that, I would suggest to just ignore it for now.

**Review by JulianFP — APPROVED** [review]
Tested it again in a VM, looks all good to me now

**Comment by RaHoni** [comment]
@therealr5 I see that you worked on the bacula module could you take a look at this PR and give me feedback on what to improve

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 3]
```diff
@@ -801,6 +801,13 @@ Make sure to also check the many updates in the [Nixpkgs library](#sec-release-2
 - `postgresql_11` has been removed since it'll stop receiving fixes on November
   9th 2023.
 
+
```
Comment (therealr5) [inline nixos/doc/manual/release-notes/rl-2311.section.md:804]: seems unrelated. Did this land in this PR by accident? :eyes:

**Review by rouven0 — COMMENTED** [review]
the diff looks good to me. But you probably want to redo this in 1 or 2 commits to conform to the clean commit style here.

Perhaps one that introduces the tls options and one that migrates `lib.mdDoc` to `mdDoc`

File: nixos/doc/manual/release-notes/rl-2311.section.md (hunk 3 above)
Comment (RaHoni) [reply nixos/doc/manual/release-notes/rl-2311.section.md:804]: Yes probably while rebasing
